### ISSUE TITLE
Automatic parsing of user-defined types

### ISSUE THREAD

**Issue by JoshuaLight** [candy-kingdom/mints#66]
_As a CLI developer, I can use inheritance instead of the `add_parser` function, which requires a tiny bit of manual work._

## Motivation
I think it'd be pretty nice to have a class like `Parsed` that would be automatically picked up by our conversion system. In simple cases, it's just a nice one step less of work to do.

## API
```py
# test.py

from mints import cli, Arg, Parsed

class Integer(Parsed):
    def __init__(self, x: str):
        self.value = int(x)

@cli
def test(x: Arg[Integer]):
    print(x.value + 1)

if __name__ == '__main__':
    cli()
```

An example of the command-line use case:
```
$ python test.py 1
2
```

**Comment by konstantin-ogulchansky** [comment]
Do we really need to inherit `Parsed`? Earlier I proposed using the following logic for parsing an arbitrary type `T`:
* If an explicit parser for `T` was specified with the `parser` decorator, then use it.
* Otherwise, use `T.__init__(x: str)`.

**Comment by JoshuaLight** [comment]
As for me, I'd prefer to be a little bit more explicit here.

Here are my main considerations:
- If I see `class Integer(Parsed)`, I'm pretty sure that the `Integer` doesn't represent an _actual_ integer. This works well with my object-oriented mind.
- If I see `class Integer`, I need to see more to be sure.

Here are my additional considerations:
- Our library is more explicit than implicit regarding types (for example, we don't support functions without annotated arguments).

But this question is not that obvious, so here are some additional comments:
- The Python language has a lot of implicit stuff, so maybe the `Parsed` thing is a little bit like C# and thus is not idiomatic.
- On the other hand, the [Zen of Python](https://www.python.org/dev/peps/pep-0020/) has a rule: _"Explicit is better than implicit"_.

What do you think?

**Comment by konstantin-ogulchansky** [comment]
First of all, I'm always hesitant about introducing a marker interface, especially when the only purpose of such an interface lies in providing some sort of explicitness. Here we use `Parsed` only to show to someone that the input string is converted somehow. However, there is no contract at all: we don't even require subclasses to have an appropriate constructor accepting a string. Of course, you could make subclasses of `Parsed` implement, for example, a `parse` method, but what's the benefit of such an approach in comparison with the existing `@cli.parse`?

Second,

> If I see class `Integer(Parsed)`, I'm pretty sure that the `Integer` doesn't represent an _actual_ integer.

In my opinion, this a problem. When I explicitly specify `dice: Arg[List[Die]]`, I want the argument `dice` exactly to be an _actual_ list of dice, not some sort of parsed things that contain the aforesaid dice.

In conclusion, I don't think that the API would suffer without `Parsed`. Moreover, it would make some things even simpler (for example, `number: Arg[int]` is straightforward: the input string is simply passed to `int.__init__(x: str)`). All the details on how parsing works could be put to the documentation in README and docstrings.

**Comment by JoshuaLight** [comment]
To sum up, if I understand correctly, you think that using `Parsed` class is not a good idea, because:
- Marker class `Parsed` doesn't provide any contract (using an explicit `parse` method is not an option).
- The classes parsed from the command line could represent actual domain entities.
- Implicitness is safe and makes some things simpler.

To be honest, some of the arguments are debatable, but I think there is no need to dive deeper. In general, I agree that the API would not suffer without `Parsed` and, as you've pointed out, it'd be even simpler.

To conclude, we need:
- Remove the `Parsed` from the example.
- Document (with the example) the implicit parsing.

Here are some open questions:
- How should we handle the fact that in the `argparse` package any _callable_ that accepts `str` and returns an instance of some type could be used as a type argument [1]? _(For example, one could write `Arg[open]`. This also means that in the `Arg[ParsedFile]` definition the `ParsedFile` could actually be a function.)_
    * Do we need to just point this out in the docs and be consistent with that in the new version of the parser backend?
    * Or it'd be better to make the requirements a bit stronger, e.g., expect only types as type arguments?

### References
[1] From [`argparse` package documentation](https://docs.python.org/3/library/argparse.html#type): _"`type=` can take any callable that takes a single string argument and returns the converted value"_.

**Comment by konstantin-ogulchansky** [comment]
I never stated that implicitness is safe. What I said is that inheriting `Parsed` does not make things explicit.

**Comment by JoshuaLight** [comment]
> But I never stated that implicitness is safe.

Sorry, my bad. The sentence was about this exact case, not in general. And by "implicitness is safe" I mean "it's ok to be implicit here".

> we could add a check that would raise an error in a case when someone tries to pass a function to Arg

I think that this is a viable solution.

So, here are our tasks for this issue:
- Add a validation (probably, [here](https://github.com/candy-kingdom/mints/blob/develop/mints/args/typed.py#L33)) for type arguments being actual types.
- Add a unit-test for `Typed` class that checks whether an exception is thrown when a function is passed as a `type` argument.
- Document the implicit parsing at the start of the [**Types**](https://github.com/candy-kingdom/mints#types) section. Probably, it'd be better to create two separate subsections: **Implicit** and **Explicit**.

**Comment by konstantin-ogulchansky** [comment]
Looks perfect!
